core/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java

```diff
@@ -141,6 +148,89 @@ public String getDetails() {
+        return Math.max(0l, delayTimeout.millis());
+    }
+
+    /**
+     * The delay in millis when delaying assigning the shard need to expire in.
+     */
+    public long getDelayAllocationExpirationIn(Settings settings, Settings indexSettings) {
+        long delayTimeout = getAllocationDelayTimeoutSetting(settings, indexSettings);
+        if (delayTimeout == 0) {
+            return 0;
+        }
+        long delta = System.currentTimeMillis() - timestamp;
+        // account for time drift, treat it as no timeout
+        if (delta < 0) {
+            return 0;
+        }
+        return delayTimeout - delta;
+    }
+
+
+    /**
+     * Returns the number of shards that are unassigned and currently being delayed.
+     */
+    public static int getNumberOfDelayedUnassigned(Settings settings, ClusterState state) {
+        int count = 0;
+        for (ShardRouting shard : state.routingTable().shardsWithState(ShardRoutingState.UNASSIGNED)) {
+            if (shard.primary() == false) {
+                IndexMetaData indexMetaData = state.metaData().index(shard.getIndex());
+                long delay = shard.unassignedInfo().getDelayAllocationExpirationIn(settings, indexMetaData.getSettings());
+                if (delay > 0) {
+                    count++;
+                }
+            }
+        }
+        return count;
+    }
+
+    /**
+     * Finds the next (closest) delay expiration setting of an unassigned shard. Returns 0 if there are none.
+     */
+    public static long findNextDelayedAllocationSetting(Settings settings, ClusterState state) {
+        long nextDelaySetting = Long.MAX_VALUE;
+        for (ShardRouting shard : state.routingTable().shardsWithState(ShardRoutingState.UNASSIGNED)) {
+            if (shard.primary() == false) {
+                IndexMetaData indexMetaData = state.metaData().index(shard.getIndex());
+                long delayTimeoutSetting = shard.unassignedInfo().getAllocationDelayTimeoutSetting(settings, indexMetaData.getSettings());
+                if (delayTimeoutSetting > 0 && delayTimeoutSetting < nextDelaySetting) {
+                    nextDelaySetting = delayTimeoutSetting;
+                }
+            }
+        }
+        return nextDelaySetting == Long.MAX_VALUE ? 0l : nextDelaySetting;
+    }
+
+
+    /**
+     * Finds the next (closest) delay expiration of an unassigned shard. Returns 0 if there are none.
+     */
+    public static long findNextDelayedAllocationIn(Settings settings, ClusterState state) {
+        long nextDelay = Long.MAX_VALUE;
+        for (ShardRouting shard : state.routingTable().shardsWithState(ShardRoutingState.UNASSIGNED)) {
+            if (shard.primary() == false) {
+                IndexMetaData indexMetaData = state.metaData().index(shard.getIndex());
+                long delayTimeoutSetting = shard.unassignedInfo().getAllocationDelayTimeoutSetting(settings, indexMetaData.getSettings());
```

Comment:
This is redundant. getDelayAllocationExpirationIn also calls getAllocationDelayTimeoutSetting()==0 and returns 0 in that case.